Thesis.org

```diff
@@ -1,11 +1,725 @@
+*** Lens
+The previously mentioned lens in line 9 is needed to stop leaking the bookkeeping for our tail call optimization, downards into our ~Item~.
+Lenses are a popular mechanism in functional programming to, on one side, restrict the available information, while allowing changes from the restricted side to change the whole, as well.
+It does this by providing two functions: Yanker and shover. The yanker grants access to parts of the whole. If parts of that whole are changed, the shover is called to marry that part again with the whole, so the update can trickle upwards again.
+Our lens is a "first lense", because it restricts access to anything but the ~first~ element. This is relevant because when state is passed around in reacl-c, it usually comes in the form of a list where the ~first~ element is the outter and the second element is the inner state.
+The code for the ~first-lens~ can be found in listing 22.
+#+CAPTION: Elements of the list are destructured into ~first~ and ~rest~, where on the yanker side ~rest~ is discarded and on the shover side ~first~ is ignored in favor of ~small~. ~small~ represents the previously passed ~first~, which now has been updated by the restricted side.
+#+begin_src clojure
+  (defn first-lens
+    ([[first & _]]
+     first)
+    ([[_ & rest] small]
+     (vec (cons small rest))))
+#+end_src
+
+Because we don't want the internal state of our ~runner~ to leak, but we do want the state surrounding the ~runner~ to be passed down, we ignore the inner state (bookkeeping) and pass down the outter state (implicitly passed state).
+Our lense is a function with multiple arities. That means it can take different amounts of parameters. Ours being of arity of two, means it has two different signatures. A signature where it gets one parameter and a signature where it takes two parameters.
+To restrict access, so when it is called from the perspective of the child, the signature with one parameter is called. That is our yanker. Here the passed parameter is destructured and everything besides the ~first~ element is ignored. That ~first~ element is then returned. On changes to the state from within our child, the signature with two parameters is called. That is our shover. The change being the second parameter, here called ~small~. We again destructure the argument but now ignore the previously named ~first~ and instead access the previously ignored ~rest~. All that is left to do is to combine them with ~cons~ and return them as a list.
+*** Implementing Tail Call Optimization
+Previously we discussed what TCO is and why we need it. Now it is time to take a deeper look at how we implement it by examining lines 3-6 and lines 11-14 in listing 21.
+Like mentioned earlier ClojureScript already provides TCO with the ~loop~ macro, unfortunately we can't take advantage of it because of the asynchronous nature of reacl-c. This is not a problem as we can use the previously explained concept of trampolines to make our own TCO.
```

Comment:
"asynchronous nature of reacl-c" is not quite correct. reacl-c works very synchronously. I know what you mean, but I'd phrase it differently for readers who are unfamiliar with reacl-c. Maybe just: `loop` and `recur` are for recursion on functions. That's now quite what we're here for, though. We need recursion on Progs. We therefore cannot simply use `loop` and `recur` as is …

Reply:
```suggestion
As mentioned earlier ClojureScript already provides TCO for functions with the ~loop~ and ~recur~ macros. We however need recursion on our ~Progs~, because of that we cannot simply use ~loop~ and ~recur~. This is not a problem though, as we can use the previously explained concept of trampolines to make our own TCO.
```
👍